File: LAB 07/220041102_T02L07_1B.cpp
#include <iostream>
#include <stack>
#include <algorithm>
#include <string>

using namespace std;

#define br cout<<endl;

struct Node {
    int data,hight;
    Node *l, *r, *p;

    Node(int d):data(d),hight(0),l(nullptr),r(nullptr),p(nullptr){}
    ~Node(){
        delete l;
        delete r;
        delete p;
    }
};

class AVLTree{
    private:
        Node *root;

        int getHight(Node *n){
            return n?n->hight:0;
        }

        int balFact(Node *n){
            return getHight(n->l)-getHight(n->r);
        }

        void updateHight(Node *n){
            while(n){
                n->hight = max(getHight(n->l),getHight(n->r))+1;
                n=n->p;
            }
        }

        void leftRotate(Node *x){
            Node *y =x->r;
            x->r = y->l;
            if(y->l)y->l->p=x;
            y->p = x->p;
            if(!y->p)root=y;
            else if(x==x->p->l) x->p->l=y;
            else x->p->r=y;
            y->l =x;
            x->p = y;

            updateHight(x);
            updateHight(y);
        }

        void rightRotate(Node *y){
            Node *x = y->l;
            y->l = x->r;
            if(x->r)x->r->p = y;
            x->p = y->p;
            if(!y->p)root=x;
            else if(y==y->p->l)y->p->l=x;
            else y->p->r=x;
            x->r = y;
            y->p = x;

            updateHight(y);
            updateHight(x);
        }

        bool balance(Node *temp){
            bool was_balanced = true;
            while(temp){
                updateHight(temp);
                int bf= balFact(temp);
                if(bf>1){
                    was_balanced = false;
                    if(balFact(temp->l)<0){
                        cout<<"Imbalance at node: "<<temp->data<<"\nLR case"<<endl;
                        leftRotate(temp->l);
                    }
                    else cout<<"Imbalance at node: "<<temp->data<<"\nLL case"<<endl;
                    rightRotate(temp);
                }
                else if(bf < -1){
                    was_balanced = false;
                    if(balFact(temp->r)>0){
                        cout<<"Imbalance at node: "<<temp->data<<"\nRL case"<<endl;
                        rightRotate(temp->r);
                    }
                    else cout<<"Imbalance at node: "<<temp->data<<"\nRR case"<<endl;
                    leftRotate(temp);
                }
                temp = temp->p;
            }
            return was_balanced;
        }

        void insert(Node * &root,int d){
            if(!root){
                root = new Node(d);
                print_tree();
                cout<<"Balanced"<<endl;
                cout<<"Root="<<root->data<<endl;
                return;
            }
            Node *temp = root, *container = new Node(d), *par=nullptr;
            
            while(temp){
                par = temp;
                if(temp->data>d)temp = temp->l;
                else if(temp->data<d) temp = temp->r;
                else{
                    cout<<"Data already exist\n";
                    return;
                }
            }

            container->p = par;
            if(d<par->data)par->l = container;
            else par->r = container;

            updateHight(container);
            print_tree();

            bool was_balanced = balance(par);
            if(was_balanced){
                cout<<"Balanced"<<endl;
                cout<<"Root="<<root->data<<endl;
            }
            else{
                cout<<"status:";
                print_tree();
                cout<<"Root="<<root->data<<endl;
            }
        }

        void inorder(Node *n){
            if(n){
                inorder(n->l);
                cout<<n->data<<"("<<balFact(n)<<") ";
                inorder(n->r);
            }
        }

        Node *find_min(Node *n){
            while(n->l) n = n->l;
            return n;
        }

        void delete_node(Node *n) {
            if (!n) return;

            Node *par = n->p;

            if (!n->l && !n->r) { 
                if (n == root) root = nullptr;
                else if (par->l == n) par->l = nullptr;
                else par->r = nullptr;
                delete n;
            } 
            else if (n->l && n->r) {
                Node *successor = find_min(n->r), *suc_par = successor->p;
                n->data = successor->data;
                
                if(successor->p->l==successor)successor->p->l=successor->r;
                else successor->p->r=successor->r;
                

                delete successor; 
                updateHight(suc_par);
            } 
            else { 
                Node *child = (n->l) ? n->l : n->r;
                child->p = par;
                if (n == root) root = child;
                else if (par->l == n) par->l = child;
                else par->r = child;
                delete n;
            }

            if (par) {
                updateHight(par);
                bool was_balanced = balance(par);
                if (was_balanced) {
                    cout << "Balanced" << endl;
                    cout << "Root=" << root->data << endl;
                } else {
                    cout << "Status:";
                    print_tree();
                }
            }
        }


    public:
        AVLTree():root(nullptr){}

        //*destructors----------------------------------------
        ~AVLTree(){
            deletetree(root);
        }

        void deletetree(Node *n){
            if(n){
                deletetree(n->l);
                deletetree(n->r);
                delete n;
            }
        }
        //*end of destructors----------------------------------

        void insert(int d){
            insert(root,d);
        }

        void print_tree(){
            inorder(root);
            br;
        }

        void delete_key(int d){
            Node *temp = root;
            while(temp && temp->data!=d)temp = temp->data>d?temp->l:temp->r;
            if(temp){
                delete_node(temp);
                print_tree();
                cout<<"Root="<<root->data<<endl;
            }
            else cout<<"Data not found"<<endl;
        }
};

int main(){
    AVLTree a;
    string val;
    while(cin>>val && val!="-1"){
        if(isalpha(val[0])){
            int d;cin>>d;
            a.delete_key(d);
        }
        else a.insert(stoi(val));
    }
    cout<<"Status:";
    a.print_tree();
    return 0;
}

File: LAB 07/220041102_T02L07_1B.py
class Node:
    def __init__(self, data):
        self.data = data
        self.height = 0
        self.left = None
        self.right = None
        self.parent = None

class AVLTree:
    def __init__(self):
        self.root = None

    def get_height(self, node):
        return node.height if node else 0

    def balance_factor(self, node):
        return self.get_height(node.left) - self.get_height(node.right)

    def update_height(self, node):
        while node:
            node.height = max(self.get_height(node.left), self.get_height(node.right)) + 1
            node = node.parent

    def left_rotate(self, x):
        y = x.right
        x.right = y.left
        if y.left:y.left.parent = x
        y.parent = x.parent
        if not y.parent:self.root = y
        elif x == x.parent.left:x.parent.left = y
        else:x.parent.right = y
        y.left = x
        x.parent = y
        self.update_height(x)
        self.update_height(y)

    def right_rotate(self, y):
        x = y.left
        y.left = x.right
        if x.right:x.right.parent = y
        x.parent = y.parent
        if not x.parent:self.root = x
        elif y == y.parent.left:y.parent.left = x
        else: y.parent.right = x
        x.right = y
        y.parent = x
        self.update_height(y)
        self.update_height(x)

    def balance(self, node):
        was_balanced = True
        while node:
            self.update_height(node)
            bf = self.balance_factor(node)
            if bf > 1:
                was_balanced = False
                if self.balance_factor(node.left) < 0:
                    print(f"Imbalance at node: {node.data}\nLR case")
                    self.left_rotate(node.left)
                else:print(f"Imbalance at node: {node.data}\nLL case")
                self.right_rotate(node)
            elif bf < -1:
                was_balanced = False
                if self.balance_factor(node.right) > 0:
                    print(f"Imbalance at node: {node.data}\nRL case")
                    self.right_rotate(node.right)
                else:print(f"Imbalance at node: {node.data}\nRR case")
                self.left_rotate(node)
            node = node.parent
        return was_balanced

    def insert(self, data):
        if not self.root:
            self.root = Node(data)
            self.print_tree()
            print("Balanced")
            print(f"Root={self.root.data}")
            return

        temp = self.root
        container = Node(data)
        parent = None

        while temp:
            parent = temp
            if data < temp.data:temp = temp.left
            elif data > temp.data:temp = temp.right
            else:
                print("Data already exists")
                return

        container.parent = parent
        if data < parent.data:parent.left = container
        else:parent.right = container

        self.update_height(container)
        self.print_tree()

        was_balanced = self.balance(parent)
        if was_balanced:
            print("Balanced")
            print(f"Root={self.root.data}")
        else:
            print("status:")
            self.print_tree()
            print(f"Root={self.root.data}")

    def inorder(self, node):
        if node:
            self.inorder(node.left)
            print(f"{node.data}({self.balance_factor(node)})", end=" ")
            self.inorder(node.right)

    def find_min(self, node):
        while node.left:node = node.left
        return node

    def delete_node(self, node):
        if not node:return

        parent = node.parent

        if not node.left and not node.right:
            if node == self.root:self.root = None
            elif parent.left == node:parent.left = None
            else:parent.right = None
            del node
        elif node.left and node.right:
            successor = self.find_min(node.right)
            par=successor.parent
            node.data = successor.data
            if successor.parent.left == successor:successor.parent.left = successor.right
            else:successor.parent.right = successor.right
            if successor.right:successor.right.parent = successor.parent
            del successor
            self.update_height(par)
        else:
            child = node.left if node.left else node.right
            child.parent = parent
            if node == self.root:self.root = child
            elif parent.left == node:parent.left = child
            else:parent.right = child
            del node
        
        self.print_tree()

        if parent:
            self.update_height(parent)
            was_balanced = self.balance(parent)
            if was_balanced:print(f"Balanced\nRoot={self.root.data}")
            else:self.print_tree()

    def delete_key(self, data):
        temp = self.root
        while temp and temp.data != data:temp = temp.left if data < temp.data else temp.right
        if temp:self.delete_node(temp)
        else:print("Data not found")

    def print_tree(self):
        self.inorder(self.root)
        print()

if __name__ == "__main__":
    avl_tree = AVLTree()
    while True:
        val = input()
        if val == "-1":break
        if val[0] in "Dd":
            val=val.split()
            avl_tree.delete_key(int(val[1]))
        else:avl_tree.insert(int(val))

    print("Status:",end=" ")
    avl_tree.print_tree()
